packit/local_project.py

```diff
@@ -562,3 +566,465 @@ def fetch(self, remote: str, refspec: Optional[str] = None, force: bool = False)
+        namespace: str = "",
+        repo_name: str = "",
+        working_dir_temporary: bool = False,
+    ):
+        self.remote = remote
+        self.git_repo = git_repo
+        self.working_dir = working_dir
+        self.ref = ref
+        self.git_project = git_project
+        self.git_service = git_service
+        self.git_url = git_url
+        self.full_name = full_name
+        self.namespace = namespace
+        self.repo_name = repo_name
+        self.working_dir_temporary = working_dir_temporary
+
+    @staticmethod
+    def from_local_project(lp: LocalProject) -> "LocalProjectCalculationState":
+        """Constructs the calculation state from an existing LocalProject."""
+        return LocalProjectCalculationState(
+            lp.remote,
+            lp.git_repo,
+            lp.working_dir,
+            lp.ref,
+            lp.git_project,
+            lp.git_service,
+            lp.git_url,
+            lp.full_name,
+            lp.namespace,
+            lp.repo_name,
+        )
+
+
+# Use an explicit sentinel class to have stricter type-checking than just object()
+class _CalculateType:
+    pass
+
+
+CALCULATE = _CalculateType()
+
+
+class LocalProjectBuilder:
+    """Class for building instances of LocalProject dynamically."""
+
+    PREREQUISITES = {
+        "git_repo": ["working_dir", "git_url"],
+        "git_project": ["git_url", "repo_name", "namespace", "git_service"],
+        "git_service": ["git_project"],
+        "ref": ["git_repo"],
+        "working_dir": ["git_repo"],
+        "git_url": ["git_project", "git_repo"],
+        "repo_name": ["git_project", "full_name"],
+        "namespace": ["git_project", "git_url", "full_name"],
+        "full_name": ["namespace", "repo_name"],
+    }
+
+    def __init__(
+        self,
+        cache: Optional[RepositoryCache] = None,
+        instances: Iterable[GitService] = None,
+        offline: bool = False,
+    ):
+        """Creates a builder instance.
+
+        Args:
+            cache: Repository cache that may be used for getting repos.
+            instances: List of GitService instances to utilise while building.
+            offline: Whether only offline operations should be performed in this builder.
+        """
+        self._instances = instances or []
+        self._cache = cache
+        self.offline = offline
+
+    def _add_prerequisites_to_calculations(self, to_calculate: Set[str]) -> None:
+        """Adds calculation prerequisites into to_calculate set.
+
+        If a caller of this class requests git_repo to be calculated, they should not
+        have to care about the prerequisites to git_repo, hence we add those to the
+        calculation set as well.
+        """
+        change = True
+        while change:
+            len_before = len(to_calculate)
+            for calc in to_calculate.copy():
+                to_calculate.update(self.PREREQUISITES.get(calc, []))
+            change = len(to_calculate) > len_before
```

Comment:
Would be really nice to log the calculation path/structure/process/dependencies.